torch/csrc/jit/python/pybind_utils.h

```diff
@@ -144,6 +144,36 @@ struct VISIBILITY_HIDDEN PythonFutureWrapper
         PyObjectType::get()));
   }
 
+  void add_done_callback(py::function cb) {
+    auto pf = std::make_shared<PythonFunctionGuard>(std::move(cb));
+    fut->addCallback(std::bind(
+        [pyFut(this->getPtr())](std::shared_ptr<PythonFunctionGuard> pf) {
+          try {
+            pybind11::gil_scoped_acquire ag;
+            pf->func_(pyFut);
+          } catch (py::error_already_set& e) {
+            {
+              pybind11::gil_scoped_acquire ag;
+              // Release ownership on py::objects and also restore Python
+              // Error Indicator.
+              e.restore();
+              // Clear the Python Error Indicator as we has recorded the
+              // exception in the response message.
+              PyErr_Clear();
+            }
+            // Log and ignore exceptions raised through the callback
```

Comment:
I might have missed some context here, but what's the reasoning behind `add_done_callback` swallowing exceptions but `then()` throwing the error? It seems like we should have consistency in the APIs here?

Reply:
I just looked through some of the old comments in the PR, and I think the reason was because of the python documentation here: https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.add_done_callback. I wrote `add_done_callback` to mimic the same behavior that the python library has, which is to log but swallow exceptions. So I guess the question is whether it's worse to (a) have inconsistencies between our `add_done_callback` and python's , or (b) have inconsistencies between our `add_done_callback` and our `then`.

Reply:
I actually did some more digging into this and I think what we have currently is fine. If you're interested I added an explanation to https://github.com/pytorch/pytorch/issues/52132